torch/csrc/jit/tensorexpr/cuda_codegen.cpp

```diff
@@ -682,176 +712,139 @@ Stmt* GPUMetaVarRewriter::mutate(const For* v) {
     }
-
-    return new_for;
   }
 
-  Stmt* mutate(const Block* v) override {
-    std::list<Stmt*> old_stmts(v->begin(), v->end());
-    std::vector<bool> need_rewrites(old_stmts.size());
-    std::vector<Stmt*> new_stmts(old_stmts.size());
-    int index = 0;
-    for (auto old_stmt : old_stmts) {
-      need_rewrite_ = false;
-      Stmt* new_stmt = Stmt::clone(old_stmt->accept_mutator(this));
-      need_rewrites[index] = need_rewrite_;
-      new_stmts[index] = new_stmt;
-      index++;
-    }
+  if (!current.empty()) {
+    innerSegments.push_back(current);
+  }
 
-    bool any_need_fix = false;
-    bool all_need_fix = need_rewrites.empty();
-    for (auto need_fix : need_rewrites) {
-      if (need_fix) {
-        any_need_fix = true;
-      } else {
-        all_need_fix = false;
+  // We are max extent in all dimensions, so need no masks at this level.
+  if (isFullRank()) {
+    // flatten inner segments.
+    std::vector<Stmt*> stmts;
+    for (auto& v : innerSegments) {
+      for (auto* s : v.stmts) {
+        stmts.push_back(s);
       }
     }
 
-    need_rewrite_ = false;
-    // If nothing needs fix, return as it is
-    if (!any_need_fix) {
-      return Block::make(new_stmts);
-    }
+    return new Block(stmts);
+  }
 
-    // If all needs fix, then we could have its parent statement to merge
-    // further. Unless the parent is a block-indx axis, then we should handle
-    // the rewrite now.
-    if (all_need_fix) {
-      Stmt* parent = v->get_parent();
-      For* loop_parent = dynamic_cast<For*>(parent);
-      if (loop_parent && loop_parent->loop_options().is_gpu_block_index()) {
-        Stmt* new_block = rewrite(new_stmts);
-        return new_block;
-      }
-      need_rewrite_ = true;
-      return Block::make(new_stmts);
+  std::vector<Stmt*> stmts;
+  int rsqi = 0;
+  for (auto& segment : innerSegments) {
+    // We never mask loops, they'll mask their contents.
+    if (segment.unmask) {
+      TORCH_INTERNAL_ASSERT(segment.stmts.size() == 1);
+      stmts.push_back(segment.stmts[0]);
+      continue;
     }
 
-    // if some needs fix, rewrites the consecutive parts
-    int start = 0;
-    int count = new_stmts.size();
-    std::vector<Stmt*> rewrite_stmts;
-    while (start < count) {
-      while (start < count && !need_rewrites[start]) {
-        rewrite_stmts.push_back(Stmt::clone(new_stmts[start]));
-        start++;
-      }
-      if (start >= count) {
-        break;
+    // If we get here, we must mask since we're not full rank and our direct
+    // child isn't a For.
+    Stmt* inner = new Block(segment.stmts);
+    // threads inside blocks.
+    auto& thread_extents = cuda_analysis_->gpu_thread_extents();
+    for (int i = 0; i < gpu_thread_vars_.size(); ++i) {
+      if (!exprEquals(current_thread_rank_[i], thread_extents[i])) {
+        // Mask it against the current rank.
+        inner = new Cond(
+            new CompareSelect(
+                gpu_thread_vars_[i],
```

Comment:
Minor: this looks good. But it is a bit silly that we are spelling the verbose version of the Cond and CompareSelect here. Maybe a good ramp-up task to add a function to simplify constructing those statements, or a TODO to remind the team for a refactoring later?

Reply:
I can write some helper constructors for this.